examples/heat-equation/heat-equation.cpp

```diff
@@ -0,0 +1,209 @@
+    auto n2 = n * n;
+    // grid point distance
+    auto h = 1.0 / n;
+    auto h2 = h * h;
+    // time step size for the simulation
+    auto tau = 1.0 / steps_per_sec;
+
+    // create an OpenMP executor
+    auto exec = gko::OmpExecutor::create();
+    // load heat source and initial state vectors
+    std::ifstream source_stream("data/source.mtx");
+    std::ifstream initial_stream("data/initial.mtx");
+    auto source = gko::read<vec>(source_stream, exec);
+    auto in_vector = gko::read<vec>(initial_stream, exec);
+    // create output vector with initial guess for
+    auto out_vector = in_vector->clone();
+    // create scalar for source update
+    auto tau_source_scalar = gko::initialize<vec>({source_scale * tau}, exec);
+    // create stencil matrix as shared_ptr for solver
+    auto stencil_matrix = gko::share(mtx::create(exec));
+    // assemble matrix
+    gko::matrix_data<> mtx_data{gko::dim<2>(n2, n2)};
+    for (int i = 0; i < n; i++) {
+        for (int j = 0; j < n; j++) {
+            auto c = i * n + j;
+            auto c_val = diffusion * tau * 4.0 / h2 + 1.0;
+            auto off_val = -diffusion * tau / h2;
+            // for each grid point: insert 5 stencil points
+            // with Dirichlet boundary conditions, i.e. with zero boundary value
+            // clang-format off
+            if (i > 0) mtx_data.nonzeros.emplace_back(c, c - n, off_val);
+            if (j > 0) mtx_data.nonzeros.emplace_back(c, c - 1, off_val);
+            mtx_data.nonzeros.emplace_back(c, c, c_val);
+            if (j < n - 1) mtx_data.nonzeros.emplace_back(c, c + 1, off_val);
+            if (i < n - 1) mtx_data.nonzeros.emplace_back(c, c + n, off_val);
+            // clang-format on
+        }
+    }
+    stencil_matrix->read(mtx_data);
+    // prepare video output
+    auto output = build_output(n, fps);
+    // build CG solver on stencil with incomplete Cholesky preconditioner
+    // stopping at 1e-10 relative accuracy
+    auto solver =
+        gko::solver::Cg<>::build()
+            .with_preconditioner(gko::preconditioner::Ic<>::build().on(exec))
+            .with_criteria(gko::stop::RelativeResidualNorm<>::build()
+                               .with_tolerance(1e-10)
+                               .on(exec))
+            .on(exec)
+            ->generate(stencil_matrix);
+    // time stamp of the last output frame (sentinel value)
+    double last_t = -t0;
+    // execute implicit Euler method: for each timestep, solve stencil system
+    for (double t = 0; t < t0; t += tau) {
+        // if enough time has passed, output the next frame
+        if (t - last_t > 1.0 / fps) {
+            last_t = t;
+            std::cout << t << std::endl;
+            output_timestep(output, n, in_vector->get_const_values());
+        }
+        // add heat
+        in_vector->add_scaled(lend(tau_source_scalar), lend(source));
+        // execute Euler step
+        solver->apply(lend(in_vector), lend(out_vector));
```

Comment:
```suggestion
        in_vector->add_scaled(gko::lend(tau_source_scalar), gko::lend(source));
        // execute Euler step
        solver->apply(gko::lend(in_vector), gko::lend(out_vector));
```

Reply:
I know we had this discussion before somewhere, but I don't really like this coding style, primarily because it complicates what it originally aimed to simplify/unify: This is more "noisy" than the inlined
```suggestion
        in_vector->add_scaled(tau_source_scalar.get(), source.get());
        // execute Euler step
        solver->apply(in_vector.get(), out_vector.get());
```
Maybe it would also make sense to revisit #179 again to work around this issue?

Reply:
I'll change this for now, but this is definitely something we can improve upon.